### PR TITLE
MM-25115- migrate post, session and status cache to cache2

### PR DESCRIPTION
**Pull request by liusy182**
#### Summary

This CR is a continuation from [#14496](https://github.com/mattermost/mattermost-server/pull/14496). 

- It migrates post, session and status cache to cache2 package, by using the new cache `Provider` interface. 

- This CR also provides a new field called `CacheProvider2` in struct `app/server`. This facilitates further migration as there are many existing usages of the `server.CacheProvider` which are based on the old cache interface.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-25115

### CODE REVIEW

**Comment by jespino** [comment]
@liusy182 I added the i18n strings :)

**Comment by liusy182** [comment]
> @liusy182 I added the i18n strings :)

got it. thanks for fixing!

File: app/post.go [hunk 1]
```diff
@@ -111,15 +112,17 @@ func (a *App) deduplicateCreatePost(post *model.Post) (foundPost *model.Post, er
 
 	// Query the cache atomically for the given pending post id, saving a record if
 	// it hasn't previously been seen.
-	value, loaded := a.Srv().seenPendingPostIdsCache.GetOrAdd(post.PendingPostId, unknownPostId, PENDING_POST_IDS_CACHE_TTL)
+	var postId string
+	errCache := a.Srv().seenPendingPostIdsCache.Get(post.PendingPostId, &postId)
```
Comment (agnivade) [inline app/post.go:116]: The convention we have been using is `nErr` for `error` types.

File: app/post.go [hunk 2]
```diff
@@ -111,15 +112,17 @@ func (a *App) deduplicateCreatePost(post *model.Post) (foundPost *model.Post, er
...
```
Comment (agnivade) [inline app/post.go:117]: Nit: let's remove the empty line.

**Comment by agnivade** [comment]
Definitely some regression in p99. But okay in average. Kinda what is expected.

Original:
original

PR:
serialized

/cc @grundleborg

**Comment by agnivade** [comment]
@lindy65  - Some testing around posts, sessions and statuses is needed. It is recommended to test in an HA setup to see if there are any edge-cases.

Look out for any sort of unexpected behavior - like getting stale data, or things not behaving correctly.

**Comment by grundleborg** [comment]
@agnivade one other question related to the load test - is there any discernable change in CPU usage on the app servers before/after this change?

**Comment by agnivade** [comment]
@grundleborg - Yes.

Original:
original


PR:
PR

It is expected because we are using gob serialization which is probably the lowest performant of encoders out there. But like we discussed in the server meeting, we decided on focus on getting this feature in first, and then focus on performance improvements. This should not be the same once we move to a better serialization algorithm, which should be done before 5.26.

**Comment by grundleborg** [comment]
OK, so about a 25% increase in CPU usage. I think we need to be careful then, as this means some servers that are "close to capacity" now will not be over capacity after this is merged.

I've no objection to it being merged at this stage, but we should have the following in place:
1. A plan for what we're going to do to reduce that CPU usage penalty *before* 5.26
2. A plan for an emergency temporary revert if, as we approach the feature freeze, we haven't got (1) sorted, so we can pull this out of 5.26 and reapply to 5.28.

**Comment by agnivade** [comment]
Totally agree.

1. The plan is to switch to a better serialization algorithm. I can do it any time.
2. I don't believe 2 will be necessary. But in case it is, I think the simplest approach would be to revert all the commits that made this change. One concern might be the code churn produced by reverting multiple commits. But it's not too bad, we are just changing one interface with another. We could also have an internal constant flag which determines whether to switch to serialized cache or not, in which case, we need to bring back the old code again.

@liusy182 - Do you have an idea as to how many more commits would be required to switch over the rest of the caches to cache2 ? I think this is the 2nd commit right ? It might be better to combine the rest in a single commit so that it becomes easy to revert later (in case).

**Comment by liusy182** [comment]
I already have the [3rd commit](https://github.com/mattermost/mattermost-server/pull/14691) and I was planning to have 10~15 smaller commits dependent on that, with each one only migrating 1 cache user.

The comments totally make sense. To summarize, we have 2 options:

1. make 1 single giant commit to do the migration so that it is easier to revert back later. It will modify 20+ files or even more.

2. have a config option to allow switching back to the old cache implementation. This will ease the urgency of doing performance optimization in the same release.

I have some thoughts on No.2. We might be able to have the `cache2.LRU` wraps `cache/lru.LRU`. Internally, `cache2.LRU` reads a config and decides whether to use the new serializable implementation or the old `cache/lru.LRU` implementation. Do we want to explore this option or No.1 ?

@agnivade @grundleborg

**Comment by agnivade** [comment]
Personally, I think 1 is fine. I don't see any issue with a big commit. It's essentially replacing interface names and changing some input params. A big commit will also help me get the full performance impact with a single commit.

5.26 release is still far away and I can take care of the performance improvement part.

**Comment by liusy182** [comment]
sounds good. I will incorporate the rest of the changes into #14691 @agnivade

**Comment by agnivade** [comment]
Awesome, thanks @liusy182 ! Excited to see the full PR.

**Pull request merged**